File: backend/app/schemas/interaction.py
"""Likes / Comments — see BACKEND_SPEC.md §3."""
from __future__ import annotations

from pydantic import Field, model_validator

from app.schemas._base import CamelModel, UtcDateTime
from app.schemas.note import NoteAuthorOut


class CommentIn(CamelModel):
    content: str = Field(min_length=1, max_length=4000)
    parent_id: str | None = None
    reply_to_sid: str | None = None
    images: list[str] = Field(default_factory=list, max_length=9)
    # Optional anchor — quote a span of the note's rendered body.
    anchor_text: str | None = Field(default=None, max_length=4000)
    anchor_offset_start: int | None = Field(default=None, ge=0)
    anchor_offset_end: int | None = Field(default=None, ge=0)

    @model_validator(mode="after")
    def validate_shape(self) -> CommentIn:
        has_start = self.anchor_offset_start is not None
        has_end = self.anchor_offset_end is not None
        if has_start != has_end:
            raise ValueError("锚点偏移必须成对提供")
        if has_start and has_end and self.anchor_offset_end <= self.anchor_offset_start:
            raise ValueError("anchorOffsetEnd 必须大于 anchorOffsetStart")
        if self.reply_to_sid is not None and self.parent_id is None:
            raise ValueError("replyToSid 仅可用于楼内回复")
        return self


class CommentOut(CamelModel):
    id: str
    note_id: str
    author: NoteAuthorOut
    content: str
    created_at: UtcDateTime
    parent_id: str | None = None
    reply_to_sid: str | None = None
    reply_to: NoteAuthorOut | None = None
    images: list[str] = Field(default_factory=list)
    status: str = "visible"
    likes: int = Field(default=0, ge=0)
    dislikes: int = Field(default=0, ge=0)
    liked_by_me: bool = False
    disliked_by_me: bool = False
    anchor_text: str | None = None
    anchor_offset_start: int | None = None
    anchor_offset_end: int | None = None


class PaginatedComments(CamelModel):
    items: list[CommentOut]
    next_cursor: str | None = None

File: frontend/src/api/schemas/interaction.ts
import { z } from 'zod'
import { NoteAuthorSchema } from './note'

/**
 * Mirror of backend `CommentOut` (see backend/app/schemas/interaction.py).
 * Anchor fields are optional — when present, the comment quotes a span of
 * the note's rendered body and the UI renders an inline blockquote.
 */
export const CommentSchema = z.object({
  id: z.string(),
  noteId: z.string(),
  author: NoteAuthorSchema,
  content: z.string(),
  createdAt: z.string(),
  /** 两层楼中楼：顶层评论 parentId=null；楼内回复 parentId=顶层评论 id。 */
  parentId: z.string().nullish(),
  replyToSid: z.string().nullish(),
  /** 被 @ 回复的用户（后端 reply_to 联查），用于「回复 @某人」。 */
  replyTo: NoteAuthorSchema.nullish(),
  /** 评论图片（0~9 张）。`.default([])` 兼容不带该字段的旧/mock 载荷。 */
  images: z.array(z.string()).default([]),
  status: z.string().default('visible'),
  likes: z.number().int().nonnegative().default(0),
  dislikes: z.number().int().nonnegative().default(0),
  likedByMe: z.boolean().default(false),
  dislikedByMe: z.boolean().default(false),
  anchorText: z.string().nullish(),
  anchorOffsetStart: z.number().int().nullish(),
  anchorOffsetEnd: z.number().int().nullish(),
})
export type Comment = z.infer<typeof CommentSchema>

export const PaginatedCommentsSchema = z.object({
  items: z.array(CommentSchema),
  nextCursor: z.string().nullable(),
})
export type PaginatedComments = z.infer<typeof PaginatedCommentsSchema>

export type CommentIn = {
  content: string
  /** 楼内回复时带上顶层评论 id + 被回复用户 sid（后端限两层）。 */
  parentId?: string | null
  replyToSid?: string | null
  /** 评论图片 URL（≤9，先经上传接口拿到 URL）。 */
  images?: string[]
  anchorText?: string | null
  anchorOffsetStart?: number | null
  anchorOffsetEnd?: number | null
}
